File: src/api/tags.py
import json
from datetime import datetime, timezone

from fastapi import APIRouter, Query, Request, Response
from sqlalchemy import func, select
from sqlalchemy.exc import IntegrityError

from src.api.schemas import TagCreate, TagListResponse, TagResponse, TagUpdate
from src.db.models import Bookmark, Tag

router = APIRouter(prefix="/api/tags", tags=["tags"])

# 10 种亮色系标签颜色，按创建顺序循环分配
TAG_COLORS: list[str] = [
    "#3B82F6",  # 蓝
    "#10B981",  # 翠绿
    "#F59E0B",  # 琥珀
    "#EF4444",  # 红
    "#8B5CF6",  # 紫
    "#EC4899",  # 粉
    "#06B6D4",  # 青
    "#F97316",  # 橙
    "#14B8A6",  # 蓝绿
    "#6366F1",  # 靛蓝
]


async def _next_tag_color(session) -> str:
    """根据当前标签总数，返回下一个循环颜色。"""
    result = await session.execute(select(func.count()).select_from(Tag))
    count: int = result.scalar_one()
    return TAG_COLORS[count % len(TAG_COLORS)]


def _tag_to_response(tag: Tag) -> TagResponse:
    return TagResponse(
        id=tag.id,
        name=tag.name,
        color=tag.color,
        icon=tag.icon,
        createdAt=tag.created_at,
        updatedAt=tag.updated_at,
    )


@router.get("", response_model=TagListResponse)
async def list_tags(
    request: Request,
    sort: str = Query(default="name"),
) -> TagListResponse:
    session_factory = request.app.state.session_factory

    sort_desc = sort.startswith("-")
    sort_key = sort[1:] if sort_desc else sort
    sort_map = {
        "name": Tag.name,
        "color": Tag.color,
        "createdAt": Tag.created_at,
        "updatedAt": Tag.updated_at,
    }
    sort_col = sort_map.get(sort_key, Tag.name)
    order_col = sort_col.desc() if sort_desc else sort_col.asc()

    async with session_factory() as session:
        result = await session.execute(select(Tag).order_by(order_col))
        tags = result.scalars().all()

    return TagListResponse(
        data=[_tag_to_response(tag) for tag in tags], total=len(tags)
    )


@router.get("/{tag_id}", response_model=None)
async def get_tag(tag_id: int, request: Request) -> TagResponse | Response:
    session_factory = request.app.state.session_factory

    async with session_factory() as session:
        result = await session.execute(select(Tag).where(Tag.id == tag_id))
        tag = result.scalar_one_or_none()

    if tag is None:
        return Response(
            content='{"type":"https://keeper.local/errors/tag-not-found","title":"标签不存在","status":404,"detail":"指定标签不存在"}',
            status_code=404,
            media_type="application/json",
        )

    return _tag_to_response(tag)


@router.post("", status_code=201, response_model=None)
async def create_tag(body: TagCreate, request: Request) -> TagResponse | Response:
    session_factory = request.app.state.session_factory

    now = datetime.now(timezone.utc).isoformat()

    async with session_factory() as session:
        color = body.color if body.color is not None else await _next_tag_color(session)
        tag = Tag(name=body.name, color=color, created_at=now, updated_at=now)
        if body.icon is not None:
            tag.icon = body.icon

        session.add(tag)
        try:
            await session.commit()
        except IntegrityError:
            await session.rollback()
            return Response(
                content='{"type":"https://keeper.local/errors/tag-name-conflict","title":"标签冲突","status":409,"detail":"标签名称已存在"}',
                status_code=409,
                media_type="application/json",
            )
        await session.refresh(tag)

    return _tag_to_response(tag)


@router.put("/{tag_id}", response_model=None)
async def update_tag(
    tag_id: int, body: TagUpdate, request: Request
) -> TagResponse | Response:
    session_factory = request.app.state.session_factory

    async with session_factory() as session:
        result = await session.execute(select(Tag).where(Tag.id == tag_id))
        tag = result.scalar_one_or_none()
        if tag is None:
            return Response(
                content='{"type":"https://keeper.local/errors/tag-not-found","title":"标签不存在","status":404,"detail":"指定标签不存在"}',
                status_code=404,
                media_type="application/json",
            )

        tag.name = body.name
        if body.color is not None:
            tag.color = body.color
        if body.icon is not None:
            tag.icon = body.icon
        tag.updated_at = datetime.now(timezone.utc).isoformat()

        try:
            await session.commit()
        except IntegrityError:
            await session.rollback()
            return Response(
                content='{"type":"https://keeper.local/errors/tag-name-conflict","title":"标签冲突","status":409,"detail":"标签名称已存在"}',
                status_code=409,
                media_type="application/json",
            )
        await session.refresh(tag)

    return _tag_to_response(tag)


@router.delete("/{tag_id}", status_code=204, response_model=None)
async def delete_tag(
    tag_id: int,
    request: Request,
    cascade: bool = Query(default=False),
) -> Response:
    session_factory = request.app.state.session_factory

    async with session_factory() as session:
        tag_result = await session.execute(select(Tag).where(Tag.id == tag_id))
        tag = tag_result.scalar_one_or_none()
        if tag is None:
            return Response(
                content='{"type":"https://keeper.local/errors/tag-not-found","title":"标签不存在","status":404,"detail":"指定标签不存在"}',
                status_code=404,
                media_type="application/json",
            )

        ref_count_r = await session.execute(
            select(func.count())
            .select_from(Bookmark)
            .where(Bookmark.tag_ids.contains(str(tag_id)))
        )
        referenced: int = ref_count_r.scalar_one()

        if not cascade and referenced > 0:
            return Response(
                content=(
                    "{"
                    '"type":"https://keeper.local/errors/tag-in-use",'
                    '"title":"标签被引用",'
                    '"status":409,'
                    f'"detail":"有 {referenced} 条书签仍在使用该标签"'
                    "}"
                ),
                status_code=409,
                media_type="application/json",
            )

        if cascade:
            affected_r = await session.execute(
                select(Bookmark).where(Bookmark.tag_ids.contains(str(tag_id)))
            )
            affected_bookmarks = affected_r.scalars().all()
            for bookmark in affected_bookmarks:
                try:
                    tag_ids_list = json.loads(bookmark.tag_ids)
                except (TypeError, ValueError):
                    tag_ids_list = []
                if isinstance(tag_ids_list, list) and tag_id in tag_ids_list:
                    bookmark.tag_ids = json.dumps(
                        [item for item in tag_ids_list if item != tag_id]
                    )

        await session.delete(tag)
        await session.commit()

    return Response(status_code=204)

File: src/handlers/tags.rs
//! 标签 API 处理器

use axum::{
    extract::{Path, Query, State},
    http::StatusCode,
    response::IntoResponse,
    Json,
};
use chrono::Utc;
use rusqlite::params;
use serde::Deserialize;

use crate::{
    db::models::Tag,
    error::{AppError, Result},
    handlers::schemas::{
        TagCreateRequest, TagListResponse, TagResponse, TagUpdateRequest,
    },
    state::AppState,
};

/// 10 种亮色系标签颜色，按创建顺序循环分配
const TAG_COLORS: &[&str] = &[
    "#3B82F6", // 蓝
    "#10B981", // 翠绿
    "#F59E0B", // 琥珀
    "#EF4444", // 红
    "#8B5CF6", // 紫
    "#EC4899", // 粉
    "#06B6D4", // 青
    "#F97316", // 橙
    "#14B8A6", // 蓝绿
    "#6366F1", // 靛蓝
];

/// 标签列表查询参数
#[derive(Debug, Deserialize)]
pub struct ListTagsQuery {
    #[serde(default = "default_sort")]
    sort: String,
}

fn default_sort() -> String {
    "name".to_string()
}

/// GET /api/tags
///
/// 获取标签列表，支持排序。
pub async fn list_tags(
    State(state): State<AppState>,
    Query(query): Query<ListTagsQuery>,
) -> Result<Json<TagListResponse>> {
    let sort = query.sort.clone();
    
    let db = state.db.clone();
    let tags: Vec<Tag> = tokio::task::spawn_blocking(move || {
        let conn = db.lock().unwrap();
        // 解析排序参数
        let sort_desc = sort.starts_with('-');
        let sort_key = if sort_desc {
            &sort[1..]
        } else {
            &sort
        };

        let order_by = match sort_key {
            "name" => "name",
            "color" => "color",
            "createdAt" => "created_at",
            "updatedAt" => "updated_at",
            _ => "name",
        };

        let sort_dir = if sort_desc { "DESC" } else { "ASC" };
        let sql = format!("SELECT * FROM tags ORDER BY {} {}", order_by, sort_dir);

        let mut stmt = conn.prepare(&sql).unwrap();
        let tag_iter = stmt.query_map([], |row| Tag::from_row(row)).unwrap();

        let mut result = Vec::new();
        for tag in tag_iter {
            result.push(tag.unwrap());
        }
        Ok::<Vec<Tag>, rusqlite::Error>(result)
    }).await
    .map_err(|e| AppError::Internal(format!("任务执行失败: {}", e)))?
    .map_err(|e| AppError::Internal(format!("数据库错误: {}", e)))?;

    let total = tags.len() as i64;
    let data: Vec<TagResponse> = tags.into_iter().map(tag_to_response).collect();

    Ok(Json(TagListResponse { data, total }))
}

/// GET /api/tags/{id}
///
/// 获取单个标签。
pub async fn get_tag(
    State(state): State<AppState>,
    Path(tag_id): Path<i64>,
) -> Result<Json<TagResponse>> {
    let db = state.db.clone();
    let tag = tokio::task::spawn_blocking(move || {
        let conn = db.lock().unwrap();
        let mut stmt = conn.prepare("SELECT * FROM tags WHERE id = ?").unwrap();
        stmt.query_row([tag_id], |row| Tag::from_row(row))
    }).await
    .map_err(|e| AppError::Internal(format!("任务执行失败: {}", e)))?
    .map_err(|_| AppError::NotFound("标签不存在".to_string()))?;

    Ok(Json(tag_to_response(tag)))
}

/// POST /api/tags
///
/// 创建标签，自动分配颜色。
pub async fn create_tag(
    State(state): State<AppState>,
    Json(req): Json<TagCreateRequest>,
) -> Result<impl IntoResponse> {
    // 验证名称
    if req.name.is_empty() {
        return Err(AppError::BadRequest("标签名称不能为空".to_string()));
    }

    // 获取当前标签数量用于颜色分配
    let db = state.db.clone();
    let count: i64 = tokio::task::spawn_blocking(move || {
        let conn = db.lock().unwrap();
        conn.query_row("SELECT COUNT(*) FROM tags", [], |row| row.get(0))
    }).await
    .map_err(|e| AppError::Internal(format!("任务执行失败: {}", e)))?
    .map_err(|e| AppError::Internal(format!("数据库错误: {}", e)))?;

    // 自动分配颜色
    let color = req.color.unwrap_or_else(|| {
        TAG_COLORS[count as usize % TAG_COLORS.len()].to_string()
    });

    let now = Utc::now();
    let icon = req.icon.unwrap_or_default();
    let name = req.name.clone();

    // 插入标签
    let db = state.db.clone();
    let result = tokio::task::spawn_blocking(move || {
        let conn = db.lock().unwrap();
        conn.execute(
            "INSERT INTO tags (name, color, icon, created_at, updated_at) VALUES (?1, ?2, ?3, ?4, ?5)",
            params![name, color, icon, now, now],
        )
    }).await;

    match result {
        Ok(Ok(_)) => {
            // 获取刚创建的标签
            let db = state.db.clone();
            let tag = tokio::task::spawn_blocking(move || {
                let conn = db.lock().unwrap();
                let mut stmt = conn.prepare("SELECT * FROM tags WHERE name = ?").unwrap();
                stmt.query_row([&req.name], |row| Tag::from_row(row))
            }).await
            .map_err(|e| AppError::Internal(format!("任务执行失败: {}", e)))?
            .map_err(|e| AppError::Internal(format!("查询标签失败: {}", e)))?;

            Ok((StatusCode::CREATED, Json(tag_to_response(tag))))
        }
        _ => Err(AppError::Conflict("标签名称已存在".to_string())),
    }
}

/// PUT /api/tags/{id}
///
/// 更新标签。
pub async fn update_tag(
    State(state): State<AppState>,
    Path(tag_id): Path<i64>,
    Json(req): Json<TagUpdateRequest>,
) -> Result<Json<TagResponse>> {
    // 检查标签是否存在
    let db = state.db.clone();
    let exists: bool = tokio::task::spawn_blocking(move || {
        let conn = db.lock().unwrap();
        let count: i64 = conn.query_row(
            "SELECT COUNT(*) FROM tags WHERE id = ?",
            [tag_id],
            |row| row.get(0),
        ).unwrap();
        Ok::<bool, rusqlite::Error>(count > 0)
    }).await
    .map_err(|e| AppError::Internal(format!("任务执行失败: {}", e)))?
    .map_err(|e| AppError::Internal(format!("数据库错误: {}", e)))?;

    if !exists {
        return Err(AppError::NotFound("标签不存在".to_string()));
    }

    let now = Utc::now();

    // 更新标签 - 空字符串表示不更新该字段
    let db = state.db.clone();
    let update_result = tokio::task::spawn_blocking(move || {
        let conn = db.lock().unwrap();
        conn.execute(
            "UPDATE tags SET 
                name = COALESCE(NULLIF(?1, ''), name), 
                color = COALESCE(NULLIF(?2, ''), color), 
                icon = COALESCE(NULLIF(?3, ''), icon), 
                updated_at = ?4 
             WHERE id = ?5",
            params![req.name, req.color, req.icon, now, tag_id],
        )
    }).await;

    match update_result {
        Ok(Ok(_)) => {
            // 获取更新后的标签
            let db = state.db.clone();
            let tag = tokio::task::spawn_blocking(move || {
                let conn = db.lock().unwrap();
                let mut stmt = conn.prepare("SELECT * FROM tags WHERE id = ?").unwrap();
                stmt.query_row([tag_id], |row| Tag::from_row(row))
            }).await
            .map_err(|e| AppError::Internal(format!("任务执行失败: {}", e)))?
            .map_err(|_| AppError::NotFound("标签不存在".to_string()))?;

            Ok(Json(tag_to_response(tag)))
        }
        _ => Err(AppError::Conflict("标签名称已存在".to_string())),
    }
}

/// DELETE /api/tags/{id}
///
/// 删除标签。
/// Query 参数: cascade (默认 false)
#[derive(Debug, Deserialize)]
pub struct DeleteTagQuery {
    #[serde(default)]
    cascade: bool,
}

pub async fn delete_tag(
    State(state): State<AppState>,
    Path(tag_id): Path<i64>,
    Query(query): Query<DeleteTagQuery>,
) -> Result<StatusCode> {
    // 检查标签是否存在
    let db = state.db.clone();
    let tag_exists: bool = tokio::task::spawn_blocking(move || {
        let conn = db.lock().unwrap();
        let count: i64 = conn.query_row(
            "SELECT COUNT(*) FROM tags WHERE id = ?",
            [tag_id],
            |row| row.get(0),
        ).unwrap();
        Ok::<bool, rusqlite::Error>(count > 0)
    }).await
    .map_err(|e| AppError::Internal(format!("任务执行失败: {}", e)))?
    .map_err(|e| AppError::Internal(format!("数据库错误: {}", e)))?;

    if !tag_exists {
        return Err(AppError::NotFound("标签不存在".to_string()));
    }

    // 检查书签引用
    if !query.cascade {
        let db = state.db.clone();
        let referenced: i64 = tokio::task::spawn_blocking(move || {
            let conn = db.lock().unwrap();
            let mut count = 0i64;
            let mut stmt = conn.prepare("SELECT tag_ids FROM bookmarks").unwrap();
            let rows = stmt.query_map([], |row| {
                let tag_ids_json: String = row.get(0)?;
                Ok(tag_ids_json)
            }).unwrap();

            for tag_ids_json in rows {
                let tag_ids_json = tag_ids_json.unwrap();
                let tag_ids: Vec<i64> = serde_json::from_str(&tag_ids_json).unwrap_or_default();
                if tag_ids.contains(&tag_id) {
                    count += 1;
                }
            }
            Ok::<i64, rusqlite::Error>(count)
        }).await
        .map_err(|e| AppError::Internal(format!("任务执行失败: {}", e)))?
        .map_err(|e| AppError::Internal(format!("检查引用失败: {}", e)))?;

        if referenced > 0 {
            return Err(AppError::Conflict(format!(
                "有 {} 条书签仍在使用该标签",
                referenced
            )));
        }
    } else {
        // 级联删除：从所有书签中移除该标签
        let db = state.db.clone();
        tokio::task::spawn_blocking(move || {
            let conn = db.lock().unwrap();
            let mut stmt = conn.prepare("SELECT id, tag_ids FROM bookmarks").unwrap();
            let rows = stmt.query_map([], |row| {
                let id: String = row.get(0)?;
                let tag_ids_json: String = row.get(1)?;
                Ok((id, tag_ids_json))
            }).unwrap();

            for row in rows {
                let (id, tag_ids_json) = row.unwrap();
                let mut tag_ids: Vec<i64> = serde_json::from_str(&tag_ids_json).unwrap_or_default();
                if tag_ids.contains(&tag_id) {
                    tag_ids.retain(|&x| x != tag_id);
                    let new_tag_ids = serde_json::to_string(&tag_ids).unwrap();
                    conn.execute(
                        "UPDATE bookmarks SET tag_ids = ? WHERE id = ?",
                        [&new_tag_ids, &id],
                    ).unwrap();
                }
            }
            Ok::<(), rusqlite::Error>(())
        }).await
        .map_err(|e| AppError::Internal(format!("任务执行失败: {}", e)))?
        .map_err(|e| AppError::Internal(format!("级联删除失败: {}", e)))?;
    }

    // 删除标签
    let db = state.db.clone();
    tokio::task::spawn_blocking(move || {
        let conn = db.lock().unwrap();
        conn.execute("DELETE FROM tags WHERE id = ?", [tag_id])
    }).await
    .map_err(|e| AppError::Internal(format!("任务执行失败: {}", e)))?
    .map_err(|e| AppError::Internal(format!("删除标签失败: {}", e)))?;

    Ok(StatusCode::NO_CONTENT)
}

/// 将 Tag 转换为 TagResponse
fn tag_to_response(tag: Tag) -> TagResponse {
    TagResponse {
        id: tag.id,
        name: tag.name,
        color: tag.color,
        icon: tag.icon,
        created_at: tag.created_at,
        updated_at: tag.updated_at,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_tag_colors() {
        assert_eq!(TAG_COLORS.len(), 10);
        assert_eq!(TAG_COLORS[0], "#3B82F6");
    }

    #[test]
    fn test_default_sort() {
        assert_eq!(default_sort(), "name");
    }
}
